### ISSUE TITLE
Add support for top level nested records in R2DBC queries

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14541]
It appears that top level nested records are simply not supported in R2DBC queries?

E.g. this test in PostgreSQL:

```java
public void testR2DBCFetchNestedRecords() throws Exception {
    assumeR2DBCAvailable();

    try {
        newBook(5).store();

        ExceptionConsumer<? super DSLContext> test = ctx -> {
            List<Record2<Integer, Record2<Integer, String>>> r2 =
            Flux.from(r2dbc(ctx)
                    .select(TBook_ID(), row(TAuthor_ID(), TAuthor_LAST_NAME()))
                    .from(TBook())
                    .leftJoin(TAuthor()).on(TBook_AUTHOR_ID().eq(TAuthor_ID()))
                    .orderBy(TBook_ID()))
                .collectList()
                .block();

            assertEquals(5, r2.size());
            assertEquals(asList(1, 2, 3, 4, 5), seq(r2).map(x -> x.value1()).toList());
            assertEquals(asList(1, 1, 2, 2, null), seq(r2).map(x -> x.value2().value1()).toList());
            assertEquals(seq(BOOK_LAST_NAMES).concat((String) null).toList(), seq(r2).map(x -> x.value2().value2()).toList());
        };

        new MultisetTests<>(delegate).testMultisetEmulations(test);
    }
    finally {
        resetBooks();
    }
}
```

Producing this query:

```sql
select
  "public"."t_book"."id", 
  row ("public"."t_author"."id", "public"."t_author"."last_name") as "nested"
from "public"."t_book" 
  left outer join "public"."t_author" 
  on "public"."t_book"."author_id" = "public"."t_author"."id"
order by "public"."t_book"."id"
```

But fails with this result, as the left join doesn't produce `null` on the last value:

```
java.lang.AssertionError: expected:<[1, 1, 2, 2, null]> but was:<[1, 1, 2, 2, 1]>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.jooq.test.util.BaseTest.assertEquals(BaseTest.java:295)
	at org.jooq.test.all.testcases.R2DBCTests.lambda$65(R2DBCTests.java:1008)
	at org.jooq.test.all.testcases.MultisetTests.lambda$469(MultisetTests.java:3081)
	at org.jooq.test.all.testcases.MultisetTests.lambda$472(MultisetTests.java:3099)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3138)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3126)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3099)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3081)
	at org.jooq.test.all.testcases.R2DBCTests.testR2DBCFetchNestedRecords(R2DBCTests.java:1012)
	at org.jooq.test.jOOQAbstractTest.testR2DBCFetchNestedRecords(jOOQAbstractTest.java:3011)
	at 
```

----
See also:
- https://stackoverflow.com/q/75175425/521799

**Comment by lukaseder** [comment]
Error when running the test on MySQL:

```
java.lang.NullPointerException: Cannot invoke "org.jooq.Record2.value1()" because the return value of "org.jooq.Record2.value2()" is null
	at org.jooq.test.all.testcases.R2DBCTests.lambda$67(R2DBCTests.java:1008)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$Itr.forEachRemaining(ArrayList.java:1003)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1921)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.jooq.lambda.SeqImpl.collect(SeqImpl.java:190)
	at org.jooq.lambda.Seq.toList(Seq.java:9201)
	at org.jooq.lambda.SeqImpl.toList(SeqImpl.java:558)
	at org.jooq.test.all.testcases.R2DBCTests.lambda$65(R2DBCTests.java:1008)
	at org.jooq.test.all.testcases.MultisetTests.lambda$469(MultisetTests.java:3081)
	at org.jooq.test.all.testcases.MultisetTests.lambda$476(MultisetTests.java:3104)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3138)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3126)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3104)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3081)
	at org.jooq.test.all.testcases.R2DBCTests.testR2DBCFetchNestedRecords(R2DBCTests.java:1012)
	at org.jooq.test.jOOQAbstractTest.testR2DBCFetchNestedRecords(jOOQAbstractTest.java:3011)
```

The feature really wasn't yet supported.

**Comment by lukaseder** [comment]
Indeed, some of the logic for flattening/unflattening top level nested records (both `ROW` and embeddables) resides in `org.jooq.impl.CursorImpl`, and is thus not available to R2DBC clients yet.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0 and 3.17.8 (https://github.com/jOOQ/jOOQ/issues/14542)
